bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/device/TapoUniversalDevice.java

```diff
@@ -0,0 +1,234 @@
+                        refreshInfo = true;
+                    }
+                    break;
+                case CHANNEL_COMMAND:
+                    String[] cmd = command.toString().split(":");
+                    if (cmd.length == 1) {
+                        connector.sendCustomQuery(cmd[0]);
+                    } else if (cmd.length == 2) {
+                        connector.sendDeviceCommand(cmd[0], cmd[1]);
+                    } else {
+                        logger.warn("({}) wrong command format '{}'", uid, command.toString());
+                    }
+                    break;
+                default:
+                    logger.warn("({}) command type '{}' not supported for channel '{}'", uid, command.toString(),
+                            channelUID.getId());
+            }
+        }
+
+        /* refreshInfo */
+        if (refreshInfo) {
+            queryDeviceInfo();
+        }
+    }
+
+    /**
+     * SET BRIGHTNESS
+     * 
+     * @param newBrightness percentage 0-100 of new brightness
+     */
+    protected void setBrightness(Integer newBrightness) {
+        /* switch off if 0 */
+        if (newBrightness == 0) {
+            connector.sendDeviceCommand(DEVICE_PROPERTY_ON, false);
+        } else {
+            HashMap<String, Object> newState = new HashMap<>();
+            newState.put(DEVICE_PROPERTY_ON, true);
+            newState.put(DEVICE_PROPERTY_BRIGHTNES, newBrightness);
+            connector.sendDeviceCommands(newState);
+        }
+    }
+
+    /**
+     * SET COLOR
+     * 
+     * @param command
+     */
+    protected void setColor(HSBType command) {
+        HashMap<String, Object> newState = new HashMap<>();
+        newState.put(DEVICE_PROPERTY_ON, true);
+        newState.put(DEVICE_PROPERTY_HUE, command.getHue());
+        newState.put(DEVICE_PROPERTY_SATURATION, command.getSaturation());
+        newState.put(DEVICE_PROPERTY_BRIGHTNES, command.getBrightness());
+        connector.sendDeviceCommands(newState);
+    }
+
+    /**
+     * SET COLORTEMP
+     * 
+     * @param colorTemp (Integer) in Kelvin
+     */
+    protected void setColorTemp(Integer colorTemp) {
+        HashMap<String, Object> newState = new HashMap<>();
+        colorTemp = limitVal(colorTemp, BULB_MIN_COLORTEMP, BULB_MAX_COLORTEMP);
+        newState.put(DEVICE_PROPERTY_ON, true);
+        newState.put(DEVICE_PROPERTY_COLORTEMP, colorTemp);
+        connector.sendDeviceCommands(newState);
+    }
+
+    /**
+     * SET DEVICE INFOs to device
+     * 
+     * @param deviceInfo
+     */
+    @Override
+    public void setDeviceInfo(TapoDeviceInfo deviceInfo) {
+        devicePropertiesChanged(deviceInfo);
+        handleConnectionState();
+    }
+
+    /**
+     * Handle full responsebody received from connector
+     * 
+     * @param responseBody
+     */
+    public void responsePasstrough(String responseBody) {
+        logger.info("({}) received response {}", uid, responseBody);
```

Comment:
Logging to info should be used sparsely e.g. a newly started component or a user file that has been loaded. This could be debug. See https://next.openhab.org/docs/developer/guidelines.html#f-logging

Reply:
"TapoUniversalDevice" is only implemented for a special "TestDevice" for devices that are not in my hand. The to the device belonged .xml will not be implemented in the normal version. Can change this to debug but then the test-user must set the log-level to debug.